main/mini-geany.c

```diff
@@ -0,0 +1,346 @@
+/*
+*   Copyright (c) 2019, Jiri Techet
+*
+*   This source code is released for free distribution under the terms of the
+*   GNU General Public License version 2 or (at your option) any later version.
+*
+*   Provides a simple application using ctags as a library and using the same
+*   set of ctags functions as the Geany editor
+*/
+
+#include "general.h"  /* must always come first */
+
+#include "types.h"
+#include "routines.h"
+#include "mio.h"
+#include "error_p.h"
+#include "writer_p.h"
+#include "parse_p.h"
+#include "options_p.h"
+#include "trashbox_p.h"
+#include "field_p.h"
+#include "xtag_p.h"
+#include "entry_p.h"
+
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+
+static int writeEntry (tagWriter *writer, MIO * mio, const tagEntryInfo *const tag);
+static void rescanFailed (tagWriter *writer, unsigned long validTagNum);
+
+/* we need to be able to provide a custom writer using which we collect the tags */
+tagWriter customWriter = {
+	.writeEntry = writeEntry,
+	.writePtagEntry = NULL,
+	.preWriteEntry = NULL,
+	.postWriteEntry = NULL,
+	.rescanFailedEntry = rescanFailed,
+	.treatFieldAsFixed = NULL,
+	.defaultFileName = "tags_file_which_should_never_appear_anywhere",
+	.private = NULL,
+	.type = WRITER_U_CTAGS /* not really but for now we must use some of the builtin types */
+};
+
+
+/* we need to be able to provide an error printer which doesn't crash Geany on error */
+static bool nofatalErrorPrinter (const errorSelection selection,
+					  const char *const format,
+					  va_list ap, void *data CTAGS_ATTR_UNUSED)
+{
+	fprintf (stderr, "%s: ", (selection & WARNING) ? "Warning: " : "Error");
+	vfprintf (stderr, format, ap);
+	if (selection & PERROR)
+#ifdef HAVE_STRERROR
+		fprintf (stderr, " : %s", strerror (errno));
+#else
+		perror (" ");
+#endif
+	fputs ("\n", stderr);
+
+	return false;
+}
+
+
+/* we need to be able to enable all kinds for all languages (some are disabled by default) */
+static void enableAllLangKinds()
+{
+	unsigned int lang;
+
+	for (lang = 0; lang < countParsers(); lang++)
+	{
+		unsigned int kindNum = countLanguageKinds(lang);
+		unsigned int kind;
+
+		for (kind = 0; kind < kindNum; kind++)
+		{
+			kindDefinition *def = getLanguageKind(lang, kind);
+			enableKind(def, true);
+		}
+	}
+}
+
+
+/* we need to be able to initialize ctags like in main() but skipping some things */
+static void ctagsInit(void)
+{
+	initDefaultTrashBox ();
+
+	setErrorPrinter (nofatalErrorPrinter, NULL);
+	setCustomTagWriter (&customWriter);
+
+	checkRegex ();
+	initFieldObjects ();
+	initXtagObjects ();
+
+	initializeParsing ();
+	initOptions ();
+
+	/* make sure all parsers are initialized */
+	initializeParser (LANG_AUTO);
+
+	/* change default value which is false */
+	enableXtag(XTAG_TAGS_GENERATED_BY_GUEST_PARSERS, true);
```

Comment:
I think we should also enable en reference tags (`XTAG_REFERENCE_TAGS`) as those are used for more and more for includes & al.